### PR TITLE
added Category model; added liquibase changelogs: creation categories…

### PR DESCRIPTION
**Pull request by YuraKhilko**
…, books_categories tables; added Set<Category> into Book model; added corresponding changes into BookDto, and CreateBookRequestDto and BookMapper; added Category service with implementation and CategoryDto with CategoryMapper; added CategoryController

### CODE REVIEW

**Review by aaananas — CHANGES_REQUESTED** [review]
image

File: src/main/java/bookstore/mapper/BookMapper.java [hunk 1]
```diff
@@ -2,13 +2,39 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.mapstruct.AfterMapping;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.mapstruct.Named;
 
-@Mapper(config = MapperConfig.class)
+@Mapper(config = MapperConfig.class, uses = CategoryServiceImpl.class)
```
Comment (aaananas) [inline src/main/java/bookstore/mapper/BookMapper.java]:
```suggestion
@Mapper(config = MapperConfig.class)
```

File: src/main/java/bookstore/mapper/BookMapper.java [hunk 2]
```diff
@@ -2,13 +2,39 @@
...
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+        Set<Long> longSet = book.getCategories().stream()
+                .map(c -> c.getId())
+                .collect(Collectors.toSet());
+        bookDto.setCategoryIds(longSet);
+    }
+
+    @Named("bookFromId")
+    default Book bookFromId(Long id) {
+        return null;
```
Comment (aaananas) [inline src/main/java/bookstore/mapper/BookMapper.java]: not implemented, I would not use service for this case

File: src/main/java/bookstore/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @JoinTable(name = "books_categories",
+            joinColumns = @JoinColumn(name = "category_id"),
+            inverseJoinColumns = @JoinColumn(name = "book_id"))
+    private Set<Book> books;
+
+    /*    public void addBook(Book book) {
+            books.add(book);
+            book.getCategories().add(this);
+        }*/
+}
```
Comment (aaananas) [inline src/main/java/bookstore/model/Category.java]: remove comments

File: src/main/java/bookstore/service/impl/CategoryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import org.mapstruct.Named;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@Named("CategoryService")
```
Comment (aaananas) [inline src/main/java/bookstore/service/impl/CategoryServiceImpl.java]:
```suggestion
```

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/model/Category.java (hunk 3 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/model/Category.java]: I've removed comment.

File: src/main/java/bookstore/mapper/BookMapper.java (hunk 1 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/mapper/BookMapper.java]: changed algorithm of categories initialization while Book create/update requests

File: src/main/java/bookstore/mapper/BookMapper.java (hunk 2 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/mapper/BookMapper.java]: removed that method

File: src/main/java/bookstore/service/impl/CategoryServiceImpl.java (hunk 4 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/service/impl/CategoryServiceImpl.java]: changed algorithm of categories initialization while Book create/update requests

**Review by aaananas — APPROVED** [review]

**Pull request merged**